### ISSUE TITLE
[Feature request] Parametrized device discovery

### ISSUE THREAD

**Issue by Sonic-Amiga** [openhab/openhab-core#1320]
Some hardware might require the user to specify some information in order to be discovered. For instance, DeviReg Smart floor thermostat can be successfully contacted only if the user specifies a one-time password.
 OpenHAB currently supports manual device search on a given binding. This needs to be extended with some properties to fill in, likely described by an XML file, similar to channels and binding config.

**Comment by kaikreuzer** [comment]
"successfully contacted" and "discovered" are two very different things.
How does discovery technically work? Isn't it using something like UPnP or mDNS?
Discovery is only meant for revealing the existence of a potentially interesting device, not for communicating with it in any way.
For a device that requires a password, the normal process is to manually create a Thing of that type and enter the required parameters. That's more or less exactly what you ask for, so I would claim that this already exists.

**Comment by Sonic-Amiga** [comment]
It's again about DeviReg Smart, and all the scenarios are described here: https://community.openhab.org/t/solved-shared-configuration-and-custom-thing-add-process/86846/11 .
 In a nutshell: there are N devices, and there is smartphone app which has already somehow set them up and knows all of them. Devices are identified by some obscure ID, which is never told to the user, only the app itself knows them. It is possible to ask the smartphone app to share this information, a one-time password is used for this process. Once the OTP has been used up it is thrown away. So there's no real sense to represent this smartphone connection as a Thing for instance. And it would look kludgy from user's point of view IMHO. I think the following flow would be good:
1. Go to Inbox
2. Press "SEARCH FOR THINGS"
3. Choose DeviReg binding
4. A field pops up asking to enter the OTP
5. The user enters the OTP, configuration is downloaded from the phone and received Things appear in the Inbox.
 Having OpenHAB to set up all thermostats from scratch in bootstrap mode would be complicated, inconvenient and not fully do the job because:
1. You would have to mess with your wi-fi config. On a working server this is somewhat painful.
2. You would have to cold reset your already set up and working thermostats. This can be very painful.

**Comment by Sonic-Amiga** [comment]
I am now studying REST API, from this point of view parameters of POST request would do the job. If the scan doesn't require anything, parameters would be empty. This would be fully backwards-compatible.

**Comment by splatch** [comment]
`DiscoveryService` is regular OSGi component thus you can inject necessary values via `ConfigurationAdmin` layer. I know it is suboptimal as parameter is going to be changes and what I propose is definitely a workaround. 
In order to get Configuration editor inside paperui you need to create avalid service configuration descriptor. Take for example:
https://github.com/openhab/openhab-core/blob/2bb5cb7f6443b531b3bda8b24ec60187b7cdd27e/bundles/org.openhab.core.ephemeris/src/main/resources/ESH-INF/config/config.xml
Above descriptor is used in this service:
https://github.com/openhab/openhab-core/blob/2bb5cb7f6443b531b3bda8b24ec60187b7cdd27e/bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java

You might notice additional properties at `@Component` annotation. In pronciple same way should work also for very specific discovery services which require extra input.

**Comment by Sonic-Amiga** [comment]
I know that i can use something, intended for different purpose, as a workaround. This ticked is a suggestion for OpenHAB 3.0.
 Actually for now i decided to go with different approach. I know that it's possible to store arbitrary files in userdata directory; so i'll write configuration downloader tool, which will download the configuration from the phone and place the resulting JSON in e.g. userdata/devireg/. Discovery service for my binding will periodically look through this directory and display contents in Inbox. This even has an advantage of persisting list of all known devices. Should the user remove a Thing for whatever reason (e. g. reconfiguring after binding upgrade), it will not be lost but come back to Inbox, so it will be possible to re-add it back without the hassle of redoing the whole pairing.
 Initial version of the tool will work from command line. Once it's done i'll come back to Community and ask for possible ways to integrate it.
 This approach can be generalized to "enable services to create and get own REST command and introduce external modules for PaperUI, so that new REST capabilities could be used:
 But this discussion definitely departs from the topic of this ticket. We can continue exploring this approach in the community: https://community.openhab.org/t/solved-shared-configuration-and-custom-thing-add-process

**Comment by cdjackson** [comment]
Currently in the ZigBee binding I have a configuration parameter that allows me to add a string with this sort of information to allow device provisioning. Again, this is not a nice solution, but it was the only way to do this when I needed it.

Looking forward, ZWave has specific requirements for this as well (these are required by the ZWave Alliance). There must be a way to configure a provisioning list - this requires the ability to add multiple devices into the list, and also requires that they can be read back, along with their provisioning status and this must be displayed in the UI.

I would propose something as an extension to the discovery service. It should have the ability to read/write/delete a list with the following information -:

* Device name (a user provided name that could mean something such as the location, or the device model - it's just for display to remind the user what they are provisioning).
* Device identifier. This would be a binding defined identification - for example for ZigBee this would be the MAC address - similar for ZWave.
* Device provisioning information. This would be a key, or whatever the binding requires to allow joining.
* Provisioning state. This would be read back from the binding to show if the device has been provisioned, or if it's not detected etc (this would be an ```enum```).

@Sonic-Amiga it would be nice to get your thoughts on this concept - does this fit your needs or do you have additional or different requirements. This is likely something I will look to implement in the near to medium term so defining the requirements and discussing amongst other users over the coming weeks would really be good.

**Comment by Sonic-Amiga** [comment]
Hi! So, you mean, that in order to be able to discover a device, its key must be listed in the provisioning info. And this info must stay there for maintaining communication. Is this correct?

**Comment by cdjackson** [comment]
Yes, that's correct. For ZigBee, there's no requirement to present this to the user, but the keys need to be known before the device can be discovered as the device can't join the network until this is done. In ZWave there is a similar system but there are additional requirements to present the provisioning list and the status to the user in the UI.

**Comment by Sonic-Amiga** [comment]
So, keys are (sometimes) not presented to the user  You say that currently you have to enter them manually. How do you obtain them?

**Comment by cdjackson** [comment]
They are made available with the device - eg printed on a box or something like that. They then need to be entered in manually or scanned with a QR reader. 

I think that part is a bit off topic and binding dependant. For now the need is to enter keys into the provisioning list and to be able to read them and try edit status back.

**Comment by Sonic-Amiga** [comment]
Not really off topic because it helps to understand how the user would interact with this information.
My case is even worse. :) Yes, there is something similar to your provisioning data in my case. But the only way to get this provisioning data is to download it from the original smartphone app using a one-time password. The user cannot enter it manually because device keys are not printed anywhere.

**Comment by splatch** [comment]
Lately I was working with multi stage discovery of BACnet infrastructure and I did it through combination of `BridgeHandler` instances and custom `ThingHandlerService` which does discovery based on parameters set on bridge. It is working as expected. In my case process goes as follows:
- discovering network
- discovering devices
- discovering objects (attributes)

**Comment by cdjackson** [comment]
Yep - this is probably similar to what I currently do in ZigBee - I use parameters to set the install codes. It works, but as it's using parameters for something they were not intended, it's not very user friendly. 

@splatch If you had a dedicated service for this, what would it look like?

Unfortunately though this will simply not work for ZWave (at least not without a lot of very hacky coding). ZWave requires that there is a provisioning list, with the provisioning state etc, and it needs to be made available to the user.

Such an interface would also be useful with ZigBee as I could avoid the misuse of "coded parameters" to enter this information.

@kaikreuzer what is your feeling about this? Is this something that you would accept as an addition in openHAB or do you consider this too binding specific?

**Comment by Sonic-Amiga** [comment]
Yesterday i got back to my binding development and dug around OH sources. I found out it''s very easy to add own REST resources. So i have implemented a rest/devireg/receive/<otp> request. This will kick the discovery process.
 Such a solution half-does the job. Half because the PaperUI doesn't know about it. For the time being i will use a shell script + curl. But perhaps it could be a good idea to introduce "UI plugins", which could add a binding-specific fragment to PaperUI (or whatever replaces it in v3). This would, i suppose, cover all specific situations.

**Comment by Sonic-Amiga** [comment]
Here is an idea what it could look like: on pressing "search for things" and selecting the binding the user sees a wizard-style page, which tells to fill in the properties and press "next". Properties and their meaning would be entirely binding-specific. For ZigBee scenario it could be provisioning info; for Devireg it would be one-time password, etc. Since the UI would be binding-specific, it could be possible to have a multi-page wizard for multi-stage discovery.

**Comment by Sonic-Amiga** [comment]
I have recently found out that a binding actually can create own UIs. See https://community.openhab.org/t/adding-custom-paperui-page .
 So far I've done that for my binding and i am now quite happy with how it works and feels. The only missing little thing is that PaperUI doesn't know about it; and "Add thing" function effectively does nothing. I think it would be a good compromise for the binding to be able to intercept functions like "Add thing" and to redirect the user to a custom interface.
 I think this would cover all possible and impossible scenarios.
 @kaikreuzer @splatch @nils @cdjackson  PING!!! Is there anyone reading this or i am just talking to myself ?

**Comment by cdjackson** [comment]
I don't think it's a great idea to require bindings to do their own thing. The ZWave binding has had to do things outside of the standard infrastructure, and it's not really nice. We should have functionality that is common and available to all.

I don't think requiring the binding to present a UI is a good idea either (sorry). The binding should (IMHO) be abstracted away from the UI - what happens when the UI changes (ie we ditch PaperUI in OH3). This would be a nightmare.

The right way to do this (IMHO) is to provide services that meet the needs ot bindings requiring such features.
